File: practice/data_structures/linked_lists/delete_duplicate_value_nodes_from_a_sorted_linked_list/delete_duplicate_value_nodes_from_a_sorted_linked_list.py
"""
 Delete duplicate nodes
 head could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None):
       self.data = data
       self.next = next_node

 return back the head of the linked list in the below method.
"""

def RemoveDuplicates(i_head):
    if not i_head:
        return i_head
    cp = i_head
    while i_head:
        if i_head.next and i_head.data == i_head.next.data:
            hn = i_head.next
            while hn and i_head.data == hn.data:
                hn = hn.next
            i_head.next = hn
        i_head = i_head.next
    return cp

File: practice/data_structures/linked_lists/delete_duplicate_value_nodes_from_a_sorted_linked_list/delete_duplicate_value_nodes_from_a_sorted_linked_list.cpp
/*
  Remove all duplicate elements from a sorted linked list
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/

Node * RemoveDuplicates(Node * i_head)
{
    if (!i_head)
    {
        return i_head;
    }
    
    Node * cp = i_head;
    while (i_head)
    {
        if (i_head->next && (i_head->data == i_head->next->data))
        {
            Node * node = i_head->next;
            while (node && (i_head->data == node->data))
            {
                node = node->next;
            }
            i_head->next = node;
        }
        i_head = i_head->next;
    }
    
    return cp;
}

